### PR TITLE
Add mode to generate canonical PIG domains

### PR DESCRIPTION
**Pull request by sadderchris**
*Issue #, if available:* N/A

*Description of changes:*
This adds two new features:
1. Adds a new `Ion` target that spits out the computed type domains in a canonical Ion form (i.e. all output domains are pure definitions, there are no permutations or transforms)
2. Adds a `--domains`/`-f` command line argument that allows you to specify a subset of domains to generate (rather than requiring all of them)

This makes it much easier to work with and post-process PIG domains outside of PIG while still leveraging PIG to do validation and type resolution.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pig/src/org/partiql/pig/cmdline/CommandLineParser.kt [hunk 1]
```diff
@@ -80,6 +82,11 @@ class CommandLineParser {
                 |      --output-file=example.txt \
                 |      --template=template.ftl 
                 |     
+                |  pig --target=identity \
```
Comment (sadderchris) [inline pig/src/org/partiql/pig/cmdline/CommandLineParser.kt:99]: Oops, this should be `--target=ion`, I'll fix this type

File: pig/src/org/partiql/pig/domain/Util.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
+package org.partiql.pig.domain.model
+
+import com.amazon.ionelement.api.IonElement
+import com.amazon.ionelement.api.ionSexpOf
+import com.amazon.ionelement.api.ionSymbol
+import com.amazon.ionelement.api.withAnnotations
+
+/*
+ * The [toIonElement] functions below generate an s-expression representation of a [TypeUniverse].
```
Comment (sadderchris) [inline pig/src/org/partiql/pig/domain/Util.kt]: `TypeUniverse` -> `TypeDomain`

**Comment by dlurton** [comment]
My first observation is:

```
cd $pojrectRoot
./gradlew assemble
cd pig/build/distributions/pig-0.5.1-SNAPSHOT/bin
./pig -u ~/projects/partiql/partiql-lang-kotlin/lang/resources/org/partiql/type-domains/partiql.ion -t kotlin -f partiql_physical --namespace junk --output-directory junk
pig: universe file: /home/ANT.AMAZON.COM/dlurton/projects/partiql/partiql-lang-kotlin/lang/resources/org/partiql/type-domains/partiql.ion
pig: parsing the universe...
pig: permuting domains...
pig: applying Kotlin pre-processing
Exception in thread "main" java.util.NoSuchElementException: Collection contains no element matching the predicate.
	at org.partiql.pig.generator.kotlin.KTypeDomainConverterKt.convertToKTypeUniverse(KTypeDomainConverter.kt:442)
	at org.partiql.pig.MainKt.generateCode(main.kt:80)
	at org.partiql.pig.MainKt.main(main.kt:53)
```

I will have more, but possibly not for a while.

**Comment by sadderchris** [comment]
Seems to have something to do with [this TODO](https://github.com/partiql/partiql-ir-generator/blob/main/pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt#L43).  I'll look into it.

**Review by dlurton — CHANGES_REQUESTED** [review]
Overall LGTM but I think we need to figure out something slightly different for how the `<src-domain>To<dest-domain>.generated.kt` files are handled when `-f` is specified.

File: pig/test/org/partiql/pig/generator/ion/GenerateIonTest.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                    (product pair n::int t::int)
+                    (product new_pair m::symbol n::int))
+                (with thing
+                    (exclude a)
+                    (include
+                        (d a::pair)
+                        (e b::symbol)))))
+        """
+
+        val expectedDomains = """
```
[the remark this answers is not in the document]
Comment (sadderchris) [reply pig/test/org/partiql/pig/generator/ion/GenerateIonTest.kt:36]: The first one contains an explicit domain permutation (`(define permuted_domain (permute_domain ...))`), the expected result contains just pure domain definitions (`(define permuted_domain (domain ...))`).  I can point this out with a comment.

File: pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt [hunk 4]
```diff
@@ -21,40 +21,37 @@
...
-
-        KTransform(
-            sourceDomainDifference = sourceDomainDifference,
-            destDomainKotlinName = destDomain.tag.snakeToPascalCase()
-        )
-    }
+    val allKTransforms = this.statements.filterIsInstance<Transform>()
+    val sourceDomains = allKTransforms.map { dt -> allTypeDomains.singleOrNull { it.tag == dt.sourceDomainTag } }
+    val destDomains = allKTransforms.map { dt -> allTypeDomains.singleOrNull { it.tag == dt.destinationDomainTag } }
+    val validKTransforms = sourceDomains.zip(destDomains)
```
[the remark this answers is not in the document]
Comment (sadderchris) [reply pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:39]: Hmmm, I think it's probably better for it to do the right thing by default :).  I'll see how hard it is to do (1).  The Kotlin generator is a lot more complicated than the other backends, so I guess I'm not surprised it wasn't this simple.  We can chat more directly once I have a better idea of how to accomplish this.

File: pig/src/org/partiql/pig/domain/Utils.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
+package org.partiql.pig.domain
```
[the remark this answers is not in the document]
Comment (RCHowell) [reply pig/src/org/partiql/pig/domain/Utils.kt:1]: I was thinking that they should probably be private and live in the only place they are used. If needed multiple places we pull them out.

**Review by RCHowell — APPROVED** [review]

File: pig/src/org/partiql/pig/domain/Utils.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (sadderchris) [reply pig/src/org/partiql/pig/domain/Utils.kt:1]: They're in the `domain` package because that's where they are in the tests ;) (this code is almost verbatim copy-pasted from [`test/org/partiql/pig/domain/Util.kt`](https://github.com/partiql/partiql-ir-generator/blob/main/pig/test/org/partiql/pig/domain/Util.kt)).  I don't really have a strong preference here though - if you want, I'll move it.

File: pig/src/org/partiql/pig/cmdline/TargetLanguage.kt [hunk 6]
```diff
@@ -18,7 +18,8 @@ package org.partiql.pig.cmdline
 import java.io.File
 
 sealed class TargetLanguage {
```
Comment (dlurton) [inline pig/src/org/partiql/pig/cmdline/TargetLanguage.kt:20]: Consider adding 

```
abstract val domains: Set<String>?
```

and `override`-ing it in base classes.

File: pig/src/org/partiql/pig/main.kt [hunk 7]
```diff
@@ -76,7 +78,7 @@ fun generateCode(command: Command.Generate) {
     when (command.target) {
```
Comment (dlurton) [inline pig/src/org/partiql/pig/main.kt:78]: If you apply the comment I made about adding an `abstract` `domains` property to the `TargetLanguage` `sealed` class, then you can move:

```
val domains = typeUniverse.computeTypeDomains().filterDomains(command.target.domains)
```

Above this line instead of having it 4 times below.

Comment (dlurton) [reply pig/src/org/partiql/pig/main.kt:78]: (I think)

File: pig/src/org/partiql/pig/domain/Utils.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (dlurton) [reply pig/src/org/partiql/pig/domain/Utils.kt:1]: Although these functions are used by more than just the new `Ion` target though: they are also used by the parser tests.

[the remark this answers is not in the document]
Comment (dlurton) [reply pig/src/org/partiql/pig/domain/Utils.kt:1]: I also notice that you've *only* moved the limited set of functions that you need for this this to the `src` dir but left the other related functions in the `test` dir.  Since they are so closely related, I would generally prefer to move them as a single unit.

File: pig/src/org/partiql/pig/domain/Utils.kt [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import com.amazon.ionelement.api.withAnnotations
+import org.partiql.pig.domain.model.DataType
+import org.partiql.pig.domain.model.NamedElement
+import org.partiql.pig.domain.model.TupleType
+import org.partiql.pig.domain.model.TypeDomain
+
+/**
+ *  Filter a [List]<[TypeDomain]> to only contain [TypeDomain]s that are present in [domains]
+ */
+fun List<TypeDomain>.filterDomains(domains: Set<String>?) =
```
Comment (dlurton) [inline pig/src/org/partiql/pig/domain/Utils.kt:15]: This one is a unrelated to the others--can you add this as a non-extension function on `TypeUniverse` (not sure that making an extension function adds much) or at least move it to `src/org.partiql.pig/domain/Utils.kt`?

File: pig/src/org/partiql/pig/domain/Utils.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (dlurton) [reply pig/src/org/partiql/pig/domain/Utils.kt:1]: Other things you could consider:

- Making these regular instance functions on their respective types (instead of extension functions)
- ... or, since there's a number of them, they aren't just "utils" anymore, put them into a file with a more specific name.

File: pig/src/org/partiql/pig/main.kt (hunk 7 above)
Comment (sadderchris) [reply pig/src/org/partiql/pig/main.kt:78]: It requires a bit more juggling for the Kotlin language (I'm not sure I like it, but I'll publish it so you can see it).

File: pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt [hunk 9]
```diff
@@ -32,29 +33,34 @@ import org.partiql.pig.domain.model.TypeUniverse
...
 
-        KTransform(
-            sourceDomainDifference = sourceDomainDifference,
-            destDomainKotlinName = destDomain.tag.snakeToPascalCase()
-        )
-    }
+    val validKTransforms = sourceDomains.zip(destDomains)
+        .filter { (sourceDomain, destDomain) -> sourceDomain != null && destDomain != null }
+        // sourceDomain and destDomain should both be non-null here, we just checked
+        .filter { (_, destDomain) -> domains?.let { destDomain!!.tag in it } ?: true }
```
Comment (dlurton) [inline pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:50]: if you use `.mapNotNull` on lines 44 and 45 instead of `.map` you can drop the filter on line 45 and the `!!`s below.

File: pig/src/org/partiql/pig/domain/Utils.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (sadderchris) [reply pig/src/org/partiql/pig/domain/Utils.kt:1]: Ah, okay, I see what these are for now.  The reason I copied them (instead of deduplicating them) is to avoid having the tests depend on implementation details (since they are asserting against a specific format).  I'll just dedup them.

File: pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt [hunk 10]
```diff
@@ -32,29 +33,34 @@ import org.partiql.pig.domain.model.TypeUniverse
...
-    }
+    val kotlinTypeDomains: List<KTypeDomain> = allTypeDomains.filterDomains(domains)
+        .map { typeDomain -> KTypeDomainConverter(typeDomain).convert() }
 
-    val allKTransforms = this.statements.filterIsInstance<Transform>().map { dt ->
-        // TODO: handle properly if one of the calls to .single fails below...
-        val sourceDomain = allTypeDomains.single { it.tag == dt.sourceDomainTag }
-        val destDomain = allTypeDomains.single { it.tag == dt.destinationDomainTag }
-        val difference = sourceDomain.computeTransform(destDomain)
+    val allKTransforms = this.statements.filterIsInstance<Transform>()
```
Comment (dlurton) [inline pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:42]: The `K` in the name of this variable probably doesn't belong, IMHO.

File: pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt (hunk 9 above)
Comment (sadderchris) [reply pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:50]: Even worse, you can end up in a situation where the `zip`ped lists won't have the same length:

```
sourceDomains [null, null, srcDomain3]
destDomains = [null, dstDomain2, dstDomain3]
```

File: pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt (hunk 10 above)
Comment (sadderchris) [reply pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:42]: This was previously called `allKTransforms`, I was just following convention, but I can change it ¯\\\_(ツ)_/¯

File: pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt [hunk 11]
```diff
@@ -32,29 +33,34 @@ import org.partiql.pig.domain.model.TypeUniverse
...
+
+            KTransform(
+                sourceDomainDifference = sourceDomainDifference,
+                destDomainKotlinName = destDomain.tag.snakeToPascalCase()
+            )
+        }
 
-    return KTypeUniverse(kotlinTypeDomains, allKTransforms)
+    return KTypeUniverse(kotlinTypeDomains, validKTransforms)
 }
```
Comment (dlurton) [inline pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:64]: Honestly, this function took me too long to figure out how it worked.  Its entirely possible I just haven't had enough coffee this morning.  However, while digging into this in my local IDE I produced this untested version which should be equivalent.  It might be easier to read.  Do with it what you will.

```Kotlin
internal fun TypeUniverse.convertToKTypeUniverse(domains: Set<String>?): KTypeUniverse {
    val allTypeDomains: List<TypeDomain> = this.computeTypeDomains()

    val kotlinTypeDomains: List<KTypeDomain> = allTypeDomains.filterDomains(domains)
        .map { typeDomain -> KTypeDomainConverter(typeDomain).convert() }

    // find all transform statements where the domain is in the `domains` set.
    // or include all transform statements if `domains` is null
    val allTransformStatements = this.statements.filterIsInstance<Transform>()
        .filter { xform -> domains?.let {  xform.destinationDomainTag in domains } ?: true }

    val validKTransforms = allTransformStatements
        .map { xform  ->
            // we should be able to assume these calls to .single wil never fail if this function is called.
            // (missing domains cause a semantic check failure)

            val sourceDomain = allTypeDomains.single { it.tag == xform.sourceDomainTag }
            val destDomain = allTypeDomains.single { it.tag == xform.destinationDomainTag }
            
            val difference = sourceDomain.computeTransform(destDomain)

            val converter = KTypeDomainConverter(difference)
            val sourceDomainDifference = converter.convert()

            KTransform(
                sourceDomainDifference = sourceDomainDifference,
                destDomainKotlinName = destDomain.tag.snakeToPascalCase()
            )
        }

    return KTypeUniverse(kotlinTypeDomains, validKTransforms)
}
```

Comment (sadderchris) [reply pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:64]: This is less efficient, since it's iterating over all type domains (twice!) inside the `map` statement, when they can be precomputed.  This is what my version does.

File: pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt (hunk 10 above)
Comment (dlurton) [reply pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:42]: Yeah, I didn't pick the best name for that variable.  Thanks.

File: pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt (hunk 11 above)
Comment (sadderchris) [reply pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:64]: Otherwise it's functionally the same

File: pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt [hunk 12]
```diff
@@ -32,29 +33,34 @@ import org.partiql.pig.domain.model.TypeUniverse
 import org.partiql.pig.util.snakeToCamelCase
 import org.partiql.pig.util.snakeToPascalCase
 
-internal fun TypeUniverse.convertToKTypeUniverse(): KTypeUniverse {
+internal fun TypeUniverse.convertToKTypeUniverse(domains: Set<String>?): KTypeUniverse {
```
Comment (dlurton) [inline pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:36]: Ok, lastly, there's some fair bit of complexity in this function.  Is it time to add tests?

**Review by dlurton — CHANGES_REQUESTED** [review]
I added a bunch of "single-comments" for some reason instead of publishing them all at once in a single review.  I guess there's a few more changes we discussed.  I think it will be good to merge once those are addressed.

File: pig/src/org/partiql/pig/domain/Utils.kt (hunk 8 above)
Comment (sadderchris) [reply pig/src/org/partiql/pig/domain/Utils.kt:15]: It was used in a couple places before, I just inlined it :D

File: pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt (hunk 11 above)
Comment (dlurton) [reply pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:64]: To be clear I don't really care what you do with that function, but, I just felt it necessary to comment on:

> This is less efficient

Efficiency is a non-goal here while readability is. 

We're talking about a function that executes in <1ms.  This function is also called once during a build that's several minutes long.

Comment (sadderchris) [reply pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:64]: Readability is subjective - I think my version is more readable since the steps are broken down.  I've incorporated a few changes though.

Comment (sadderchris) [reply pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:64]: If you have some readability guidelines for this repo, can you point me to them?

Comment (dlurton) [reply pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt:64]: There are no readability guidelines here.  Of course readability is subjective.  That's why I don't honestly care what you do with the alternative function I posted above.  (Of course, efficiency is a concern for the generated code.)
